### PR TITLE
WARC lookup for remote zipfile (WACZ)

### LINKED ISSUE #21: Improve performance of `get_warc_from_cdxj_record`
In this function we currently re-open the WACZ each time we request a WARC record. When using a cloud provider we keep fetching the file. Also when not using a cloud provider we should not need to re-open this file every time.

### PR DESCRIPTION
**Pull request by leewesleyv**
Resolves #21 

To account for large zipfiles (WACZ) we are implementing lookup of the index/WARC record using range requests for remote files.

- [x] Add a storage handler class that abstracts logic to allow handling local and remote zipfiles
- [x] Implement logic for requesting partial zip data and parsing the data low level instead of using Python's build-in zipfile module (does not seem to be compatible without having the entire file).
- [x] Add unit-tests

**Implement different storages handlers**

- [x] Implement local storage handler class
- [x] Implement S3 storage handler class
~~Implement GC storage handler class~~
~~Implement Azure storage handler class~~
~~Implement FTP storage handler class~~

**Useful links**

- The ZIP specification https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT 
- Interesting blog article https://blog.yaakov.online/zip64-go-big-or-go-home/

### CODE REVIEW

File: scrapy_webarchive/wacz/storages.py [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+                with gzip.open(file_obj) as gz_file:
+                    return gz_file.read()
+
+            return file_obj.read()
+
+    def fetch_file(self, file_name: str) -> bytes:
+        if file_name not in self.zip_file.namelist():
+            raise FileNotFoundError(f"File {file_name} not found in ZIP archive")
+
+        with self.zip_file.open(file_name) as file_obj:
```
Comment (wvengen) [inline scrapy_webarchive/wacz/storages.py:72]: Are we opening this each time we're getting a file? A possible later improvement would be to open it once, then use the opened file.

File: scrapy_webarchive/wacz/storages.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+    def get_handler(uri: str, settings) -> ZipStorageHandler:
+        if uri.startswith('s3://'):
+            handler = S3ZipStorageHandler
+            handler.AWS_ACCESS_KEY_ID = settings["AWS_ACCESS_KEY_ID"]
+            handler.AWS_SECRET_ACCESS_KEY = settings["AWS_SECRET_ACCESS_KEY"]
+            handler.AWS_SESSION_TOKEN = settings["AWS_SESSION_TOKEN"]
+            handler.AWS_ENDPOINT_URL = settings["AWS_ENDPOINT_URL"]
+            handler.AWS_REGION_NAME = settings["AWS_REGION_NAME"]
+            handler.AWS_USE_SSL = settings["AWS_USE_SSL"]
+            handler.AWS_VERIFY = settings["AWS_VERIFY"]
```
Comment (wvengen) [inline scrapy_webarchive/wacz/storages.py:226]: There is also a setting related to whether to build the final S3 endpoint URL with the bucket name in it or not (path style), that is quite important to access non-AWS S3 endpoints.

**Review by wvengen — COMMENTED** [review]
Wow, impressive you got this, with a streaming zip-reading implementation! Looking good, at first glance.

I'm still a bit surprised that there is no existing streaming-zip-over-http implementation, but it happens that sometimes some real engineering is still necessary :) :rocket:

**Comment by wvengen** [comment]
I would say, GC/Azure/FTP are nice to haves - so feel free to add them, but I would not spend a long long time if you come across blocking issues.

File: scrapy_webarchive/wacz/storages.py (hunk 2 above)
Comment (leewesleyv) [reply scrapy_webarchive/wacz/storages.py:226]: Is this an existing setting in Scrapy? As far as I can see all the existing settings are passed on when creating the s3 client. From your comment I would guess you are referring to `AWS_ENDPOINT_URL` which we also do implement. Am I missing something here?

Comment (wvengen) [reply scrapy_webarchive/wacz/storages.py:226]: Ah, I see we don't need this setting - apparently the default path style is fine. Nothing to do here then!

File: scrapy_webarchive/wacz/storages.py (hunk 1 above)
Comment (leewesleyv) [reply scrapy_webarchive/wacz/storages.py:72]: Indeed a good suggestion! Will create a ticket for this improvement.

**Pull request merged**